lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -559,6 +559,159 @@ namespace Js
     }
 
+    // Promise.prototype.finally as described in the draft of ES 2018 Section 25.4.5.3
+    Var JavascriptPromise::EntryFinally(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+
+        ScriptContext* scriptContext = function->GetScriptContext();
+
+        AUTO_TAG_NATIVE_LIBRARY_ENTRY(function, callInfo, _u("Promise.prototype.finally"));
+        //1. Let promise be the this value
+        //2. If Type(promise) is not Object, throw a TypeError exception
+        if (args.Info.Count < 1 || !JavascriptOperators::IsObject(args[0]))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_This_NeedObject, _u("Promise.prototype.finally"));
+        }
+
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(args[0]);
+        //3. Let C be ? SpeciesConstructor(promise, %Promise%).
+        RecyclableObject* constructor = JavascriptOperators::SpeciesConstructor(promise, scriptContext->GetLibrary()->GetPromiseConstructor(), scriptContext);
+        //4. Assert IsConstructor(C)
+        Assert(JavascriptOperators::IsConstructor(constructor));
+
+        //5. If IsCallable(onFinally) is false
+        //  a. Let thenFinally be onFinally
+        //  b. Let catchFinally be onFinally 
+        //6. Else,
+        //  a. Let thenFinally be a new built-in function object as defined in ThenFinally Function.
+        //  b. Let catchFinally be a new built-in function object as defined in CatchFinally Function.
+        //  c. Set thenFinally and catchFinally's [[Constructor]] internal slots to C.
+        //  d. Set thenFinally and catchFinally's [[OnFinally]] internal slots to onFinally.
+
+        Var thenFinally = nullptr;
+        Var catchFinally = nullptr;
+
+        if (args.Info.Count > 1)
+        {
+            if (JavascriptConversion::IsCallable(args[1]))
+            {
+                //note to avoid duplicating code the ThenFinallyFunction works as both thenFinally and catchFinally using a flag
+                thenFinally = library->CreatePromiseThenFinallyFunction(EntryThenFinallyFunction, RecyclableObject::FromVar(args[1]), constructor, false);
+                catchFinally = library->CreatePromiseThenFinallyFunction(EntryThenFinallyFunction, RecyclableObject::FromVar(args[1]), constructor, true);
+            }
+            else
+            {
+                thenFinally = args[1];
+                catchFinally = args[1];
+            }
+        }
+        else
+        {
+            thenFinally = library->GetUndefined();
+            catchFinally = library->GetUndefined();
+        }
+
+        Assert(thenFinally != nullptr && catchFinally != nullptr);
+
+        //7. Return ? Invoke(promise, "then", « thenFinally, catchFinally »).
```

Comment:
>« thenFinally, catchFinally » [](start = 46, length = 29)

Please remove non-ascii characters from comments (this has caused problems when compiling on systems in some non-English locales in the past). You could replace these with `<<` and `>>` for example.

Reply:
We should have a check for this. Added a task: [https://github.com/Microsoft/ChakraCore/issues/4672](https://github.com/Microsoft/ChakraCore/issues/4672)

---
In reply to: [167734493](https://github.com/Microsoft/ChakraCore/pull/4668#discussion_r167734493) [](ancestors = 167734493)

Reply:
Will do, sorry about this, I'd copied and pasted these lines from the spec

Reply:
No worries, common gotcha.